AG-9014 Error Bar improvements for Bar Charts

olegat opened the pull request:
https://ag-grid.atlassian.net/browse/AG-9014

Use the `Series.contextNodeData` to get the `midPoint` rendering value. This makes the `errorBar.ts` file render the error bars in the centre of the bars in Bar Charts.

Also flip the X/Y keys in horizontal Bar Charts so that horizontal whiskers are used. The vertical whiskers are now optional (which I suppose makes sense, it would also allow scatter series with confidence intervals in just 1 axis).

olegat committed:
AG-9014 Position errorbar using the band-scale in bar-series

This significantly refactors the errorBar.ts file. The ErrorBars class
now uses an abstract CartesianInterface class which behaves almost
exactly the same for line, scatter and bar series (but with some
overrides for bar-series). This cuts back on the number of if-checks
that the ErrorBars class need (because it doesn't really care about the
underlying series type).

With this change, the errorbars in bar-series are centered on the top
left corner of the bars for each series. This is still wrong (it should
be in the top center of the bar), but this is good progress.

olegat committed:
AG-9014 Use public props/methods in BarInterface.convert

olegat committed:
AG-9014 Implement BarInterface.convert for horizontal bars

olegat committed:
AG-9014 Add test for errorbars on horizontal bar charts

olegat committed:
AG-9014 Simplify errorBar.ts significantly

Using Series.contextNodeData grants access to the X/Y midpoint of the
rendered node, which is all that errorBar.ts really needs.

Also, flip the X & Y error-datum keys in horizontal bar series so that
the horizontal whiskers are used.

olegat committed:
AG-9014 Add horizontal bar-chart errorbar snapshot

alantreadway reviewed (approved):
LGTM - thanks for reworking to use nodeData, looks much neater 👍

olegat committed:
AG-9014 Fix bug with Y axis on horizontal barcharts with errorbars

olegat committed:
AG-9014 Add more errorbar tests and update snapshots

olegat commented on packages/ag-charts-enterprise/src/features/error-bar/errorBar.ts:
```diff
@@ -182,104 +181,79 @@ export class ErrorBars
-    }
+    private shouldFlipXY(): boolean {
+        if (this.cartesianSeries.type === 'bar') {
+            return (this.cartesianSeries as unknown as _ModuleSupport.BarSeries).direction === 'horizontal';
```
I'm not sure what's the best way to check if the bar series is horizontal. In theory, this `as unknown as BarSeries` should in theory be safe because the `type` property is being checked. As said, they may be a cleaner way to happen this edge case.

Alternatives could be:

- Using `instanceof` instead of `as`
- Providing the `direction` value as part of the module context.

Please let me know your thoughts.

olegat replied:
Actually I've move `shouldFlipXY()` has part of `CartesianSeries` (and `BarSeries` overrides it). I find that this a much cleaner approach than having _errorBar.ts_ dig into the implementation details of the underlying series.

alantreadway replied:
SGTM - thanks!

olegat committed:
AG-9014 Avoid use of `as`

Adding a virtual Cartesian.shouldFlipXY() function which BarSeries can
override is a cleaner approach, because the errorBar.ts no longer needs
to know anything specific about the underlying series (aside from the
validation in the constructor).
